### PR TITLE
feat(contracts): add contracts about `Multi sig based proxy upgrade`

### PR DESCRIPTION
**Pull request by kangsorang**
# Description

Contracts and related code have been written for the Multi sig based proxy upgrade.
A contract upgrade is only possible if the vote following the proposal passes.
Voting privileges are granted based on a token for this purpose, which in this case is security council members.
If a proposal is voted on and passed, it will be executed after being delayed by a set timelock.
The delay in upgrading by Timelock will be more than a month.
(https://ethereum-magicians.org/t/proposed-milestones-for-rollups-taking-off-training-wheels/11571)

*NOTE:* Additional audits are underway and feedback will be adjusted.

### CODE REVIEW

**Review by Pangssu — COMMENTED** [review]
Why not separate OpenZeppelin version updates from code changes commits?

**Comment by Pangssu** [comment]
Can you change the portalGuardian address to the KromaGoverner address? If you can, you should modify `layer_one.go` and delete the portalGuardian value from the deploy-config json file. This will resolve the KROMA-022 audit issue.

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 1]
```diff
@@ -30,9 +30,14 @@ import { Types } from "../libraries/Types.sol";
 import { KromaMintableERC20 } from "../universal/KromaMintableERC20.sol";
 import { KromaMintableERC20Factory } from "../universal/KromaMintableERC20Factory.sol";
 import { KromaMintableERC721Factory } from "../universal/KromaMintableERC721Factory.sol";
+import {SecurityCouncilToken} from "../governance/SecurityCouncilToken.sol";
+import {TimeLock} from "../governance/TimeLock.sol";
+import {UpgradeGovernor} from "../governance/UpgradeGovernor.sol";
```
Comment (Pangssu) [inline packages/contracts/contracts/test/CommonTest.t.sol:35]:
```suggestion
import { SecurityCouncilToken } from "../governance/SecurityCouncilToken.sol";
import { TimeLock } from "../governance/TimeLock.sol";
import { UpgradeGovernor } from "../governance/UpgradeGovernor.sol";
```

File: packages/contracts/contracts/governance/SecurityCouncilToken.sol [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+     * @notice Initializer.
+     *
+     * @param _owner           Owner of this token contract.
+     */
+    function initialize(address _owner) initializer public {
+        __KromaSoulBoundERC721_init("KromaSecurityCouncil", "KSC", _owner);
+    }
+
+    function _baseURI() internal pure override returns (string memory) {
+        return "";
```
Comment (0xHansLee) [inline packages/contracts/contracts/governance/SecurityCouncilToken.sol:35]: If this would be base URI for security council SBT, please mark TODO for future use.

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 3]
```diff
@@ -30,9 +30,14 @@ import { Types } from "../libraries/Types.sol";
...
 import { Proxy } from "../universal/Proxy.sol";
+import { ProxyAdmin } from "../universal/ProxyAdmin.sol";
 import { AddressAliasHelper } from "../vendor/AddressAliasHelper.sol";
 
+
```
Comment (0xHansLee) [inline packages/contracts/contracts/test/CommonTest.t.sol:40]:
```suggestion
```

File: packages/contracts/package.json [hunk 4]
```diff
@@ -53,8 +53,8 @@
   },
   "dependencies": {
     "@kroma-network/core-utils": "^0.1.0",
-    "@openzeppelin/contracts": "4.7.3",
-    "@openzeppelin/contracts-upgradeable": "4.7.3",
+    "@openzeppelin/contracts": "4.9.3",
+    "@openzeppelin/contracts-upgradeable": "4.9.3",
```
Comment (0xHansLee) [inline packages/contracts/package.json:57]: I think there was a reason why you upgraded the version of contracts.
Please leave a comment for this!

File: utils/chain-ops/genesis/config.go [hunk 5]
```diff
@@ -117,6 +119,17 @@ type DeployConfig struct {
 	EIP1559Denominator uint64 `json:"eip1559Denominator"`
 
 	FundDevAccounts bool `json:"fundDevAccounts"`
+
+	// The initial value of the voting delay(unit:block)
+	VotingDelayBlocks uint64 `json:"votingDelayBlocks"`
+	// The initial value of the voting period(unit:block)
+	VotingPeriodBlocks uint64 `json:"votingPeriodBlocks"`
+	// The initial value of the proposal threshold(unit:block)
```
Comment (0xHansLee) [inline utils/chain-ops/genesis/config.go:127]: The unit of this may not be block, but the number of tokens, right?
```suggestion
	// The initial value of the proposal threshold
```

File: utils/chain-ops/genesis/testdata/test-deploy-config-devnet-l1.json [hunk 6]
```diff
@@ -38,6 +38,13 @@
   "colosseumSegmentsLengths": "3,11",
   "securityCouncilNumConfirmationRequired": 1,
   "securityCouncilOwners": [
-    "0x007D7e4391DCFdE47Dd7FD0B8E16091C5e0E1C7f","0x2A06Af6a4F325B1e0095a8AbD32888e0AbdCD04D","0xed92E7C40348A552822847384D722A6adB9AFeFA"
-  ]
+    "0x007D7e4391DCFdE47Dd7FD0B8E16091C5e0E1C7f",
+    "0x2A06Af6a4F325B1e0095a8AbD32888e0AbdCD04D",
+    "0xed92E7C40348A552822847384D722A6adB9AFeFA"
+  ],
+  "votingDelayBlocks": 0,
```
Comment (0xHansLee) [inline utils/chain-ops/genesis/testdata/test-deploy-config-devnet-l1.json:45]: We don't need token owner of security council SBT here?

File: e2e/setup.go [hunk 7]
```diff
@@ -149,6 +149,12 @@ func DefaultSystemConfig(t *testing.T) SystemConfig {
 		EIP1559Denominator: 8,
 
 		FundDevAccounts: true,
+
+		VotingDelayBlocks:               0,
```
Comment (0xHansLee) [inline e2e/setup.go:153]: Don't we need to set the owner of security council SBT?

File: packages/contracts/deploy/013-SecurityCouncilToken.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  const artifact = await hre.deployments.get('SecurityCouncilToken')
+  const token = new ethers.Contract(
+    securityCouncilTokenProxyAddress,
+    artifact.abi,
+    hre.ethers.provider.getSigner(deployer)
+  )
+  for (const guardian of hre.deployConfig.securityCouncilOwners) {
+    const balance = await token.balanceOf(guardian)
+    if (balance === 0) {
+      await token.safeMint(guardian, '')
```
Comment (0xHansLee) [inline packages/contracts/deploy/013-SecurityCouncilToken.ts:38]: Please mark TODO here since we need to provide URI to each token.

File: packages/contracts/deploy/015-UpgradeGovernor.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    hre,
+    'SecurityCouncilTokenProxy'
+  )
+  const timeLockProxyAddress = await getDeploymentAddress(
+    hre,
+    'TimeLockProxy'
+  )
+
+  await deploy(hre, 'UpgradeGovernor', {
+    args: [],
```
Comment (0xHansLee) [inline packages/contracts/deploy/015-UpgradeGovernor.ts:21]: If no arguments are required, we can delete it as done in other deploy scripts.

File: packages/contracts/contracts/test/UpgradeGovernor.t.sol [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        super.setUp();
+        _roll(1);
+        // minting to guardians
+        vm.prank(owner);
+        securityCouncilToken.safeMint(guardian1, baseUri);
+        vm.prank(owner);
+        securityCouncilToken.safeMint(guardian2, baseUri);
+
+        vm.prank(guardian1);
+        securityCouncilToken.delegate(guardian1);
```
Comment (0xHansLee) [inline packages/contracts/contracts/test/UpgradeGovernor.t.sol:30]: IMHO, there is no need to delegate token to guardian1 since the token is already minted to guardian1.
Why is this required?

**Review by 0xHansLee — COMMENTED** [review]
I would like you to provide the reference contracts that you referred to while working.

File: packages/contracts/contracts/universal/Semver.sol [hunk 11]
```diff
@@ -43,7 +43,7 @@ contract Semver {
      *
      * @return Semver contract version as a string.
      */
-    function version() public view returns (string memory) {
+    function version() public virtual view returns (string memory) {
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply packages/contracts/contracts/universal/Semver.sol:46]: The overriding is necessary because the GovernorUpgradeable contract has the same version function.
This is a choice to use the functionality of semver without modifying the GovernorUpgradeable contract.

File: packages/contracts/contracts/universal/KromaSoulBoundERC721.sol [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+     * @param tokenId The identifier for an SBT.
+     */
+    function locked(uint256 tokenId) external view returns (bool);
+}
+
+abstract contract KromaSoulBoundERC721 is
+    Initializable,
+    IERC5192,
+    ERC721Upgradeable,
+    ERC721EnumerableUpgradeable,
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply packages/contracts/contracts/universal/KromaSoulBoundERC721.sol:46]: By solidity's linearization inheritance, multiple inheritances of ERC721Upgradeable behave the same as a single inheritance.

File: packages/contracts/package.json (hunk 4 above)
Comment (kangsorang) [reply packages/contracts/package.json:57]: From 4.7.3 to 4.9.3, there have been many changes to contracts such as Timelock, governor, erc721, etc. There are also breaking changes, so we have updated to the latest release version.
(https://github.com/OpenZeppelin/openzeppelin-contracts/releases)

**Review by 0xHansLee — APPROVED** [review]
lgtm

File: utils/chain-ops/genesis/config.go [hunk 13]
```diff
@@ -117,6 +119,17 @@ type DeployConfig struct {
...
+	// The initial value of the voting delay(unit:block)
+	VotingDelayBlocks uint64 `json:"votingDelayBlocks"`
+	// The initial value of the voting period(unit:block)
+	VotingPeriodBlocks uint64 `json:"votingPeriodBlocks"`
+	// The initial value of the proposal threshold(unit:token)
+	ProposalThresholdTokens uint64 `json:"proposalThresholdTokens"`
+	// The initial value of the votes quorum fraction(unit:percent)
+	VotesQuorumFractionPercent uint64 `json:"votesQuorumFractionPercent"`
+	// The latency value of the proposal executing(unit:second)
+	ProposalExecutingLatencySeconds uint64 `json:"proposalExecutingLatencySeconds"`
```
Comment (seolaoh) [inline utils/chain-ops/genesis/config.go:132]: Don't we need to add these variables to `Check()` below?

File: utils/chain-ops/genesis/config.go [hunk 14]
```diff
@@ -90,6 +90,8 @@ type DeployConfig struct {
 	FinalSystemOwner common.Address `json:"finalSystemOwner"`
 	// GUARDIAN account in the KromaPortal
 	PortalGuardian common.Address `json:"portalGuardian"`
+	// Owner of the Guardian Token
+	SecurityCouncilTokenOwner common.Address `json:"securityCouncilTokenOwner"`
```
Comment (seolaoh) [inline utils/chain-ops/genesis/config.go:94]: Don't we need to add this variable to `InitDeveloperDeployedAddresses()` below?

File: utils/chain-ops/genesis/testdata/test-deploy-config-devnet-l1.json [hunk 15]
```diff
@@ -38,6 +38,13 @@
...
+  "votingPeriodBlocks": 25,
+  "proposalThresholdTokens": 1,
+  "votesQuorumFractionPercent": 51,
+  "proposalExecutingLatencySeconds": 1
```
Comment (seolaoh) [inline utils/chain-ops/genesis/testdata/test-deploy-config-devnet-l1.json:49]: Do we need these variables in this file?

File: e2e/setup.go [hunk 16]
```diff
@@ -149,6 +149,12 @@ func DefaultSystemConfig(t *testing.T) SystemConfig {
 		EIP1559Denominator: 8,
 
 		FundDevAccounts: true,
+
```
Comment (seolaoh) [inline e2e/setup.go:152]: Don't we need to set here like below?
```go
		SecurityCouncilTokenOwner: addresses.Deployer,
```

**Comment by seolaoh** [comment]
We need to name the branch name starting with `feat`, ..., please follow it next time :)
Also IMO, the first commit type would be better to be `chore(contracts)`, and update snapshots using `make contracts-snapshot`!

File: packages/contracts/contracts/universal/KromaSoulBoundERC721.sol [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+import "@openzeppelin/contracts-upgradeable/utils/cryptography/draft-EIP712Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/draft-ERC721VotesUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
+
+/**
+ * @title IERC5192
+ * @notice Interface for contracts that are compatible with the ERC721 standard.
+ */
+interface IERC5192 {
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply packages/contracts/contracts/universal/KromaSoulBoundERC721.sol:19]: This was a tough decision. We decided to implement the interface by embedding it in other contracts, such as ProxyAdmin.sol.

File: utils/chain-ops/genesis/config.go (hunk 14 above)
Comment (kangsorang) [reply utils/chain-ops/genesis/config.go:94]: Yes, this value is included in the deploy-config.

File: utils/chain-ops/genesis/testdata/test-deploy-config-devnet-l1.json (hunk 15 above)
Comment (kangsorang) [reply utils/chain-ops/genesis/testdata/test-deploy-config-devnet-l1.json:49]: Yes, the contracts are covered by the forge test. I'll remove this part since it's not used.

File: e2e/setup.go (hunk 16 above)
Comment (kangsorang) [reply e2e/setup.go:152]: We don't need it here.

File: e2e/e2eutils/setup.go [hunk 18]
```diff
@@ -120,6 +121,12 @@ func MakeDeployParams(t require.TestingT, tp *TestParams) *DeployParams {
 		EIP1559Denominator: 50,
 
 		FundDevAccounts: false,
+
+		VotingDelayBlocks:               0,
+		VotingPeriodBlocks:              25,
+		ProposalThresholdTokens:         1,
+		VotesQuorumFractionPercent:      51,
+		ProposalExecutingLatencySeconds: 1,
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply e2e/e2eutils/setup.go:129]: That's nice, but I think the name would be too long.

File: e2e/e2eutils/setup.go [hunk 19]
```diff
@@ -120,6 +121,12 @@ func MakeDeployParams(t require.TestingT, tp *TestParams) *DeployParams {
 		EIP1559Denominator: 50,
 
 		FundDevAccounts: false,
+
+		VotingDelayBlocks:               0,
+		VotingPeriodBlocks:              25,
+		ProposalThresholdTokens:         1,
+		VotesQuorumFractionPercent:      51,
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply e2e/e2eutils/setup.go:128]: That's nice, but I think the name would be too long...

[the remark this answers is not in the document]
Comment (seolaoh) [reply e2e/e2eutils/setup.go:128]: Yes I agree, please change it with another nice naming :)

File: bindings/Makefile [hunk 20]
```diff
@@ -6,13 +6,15 @@ all: version mkdir bindings more
 
 bindings: basefee-vault-bindings \
 		colosseum-bindings \
-    security-council-bindings \
-		erc20-bindings \
+    erc20-bindings \
```
Comment (Pangssu) [inline bindings/Makefile:9]: two tabs

**Review by seolaoh — APPROVED** [review]
LGTM

File: bindings/Makefile [hunk 21]
```diff
@@ -46,9 +49,12 @@ more:
 		-out ./bindings \
 		-contracts \
 		Colosseum,\
+UpgradeGovernor,\
 KromaMintableERC20Factory,\
 KromaMintableERC721Factory,\
+SecurityCouncilToken,\
 KromaPortal,\
+TimeLock,\
```
Comment (Pangssu) [inline bindings/Makefile:57]: ascending

File: bindings/Makefile [hunk 22]
```diff
@@ -100,6 +106,15 @@ kroma-mintable-erc721-factory-bindings: compile
...
 	bash ./gen_bindings.sh contracts/L1/KromaPortal.sol:KromaPortal $(pkg)
 
+upgrade-governor-bindings: compile
+	bash ./gen_bindings.sh contracts/governance/UpgradeGovernor.sol:UpgradeGovernor $(pkg)
+
+timelock-bindings: compile
+	bash ./gen_bindings.sh contracts/governance/TimeLock.sol:TimeLock $(pkg)
+
+security-council-token-bindings: compile
+	bash ./gen_bindings.sh contracts/governance/SecurityCouncilToken.sol:SecurityCouncilToken $(pkg)
```
Comment (Pangssu) [inline bindings/Makefile:116]: ascending

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 23]
```diff
@@ -30,7 +30,11 @@ import { Types } from "../libraries/Types.sol";
 import { KromaMintableERC20 } from "../universal/KromaMintableERC20.sol";
 import { KromaMintableERC20Factory } from "../universal/KromaMintableERC20Factory.sol";
 import { KromaMintableERC721Factory } from "../universal/KromaMintableERC721Factory.sol";
+import { SecurityCouncilToken } from "../governance/SecurityCouncilToken.sol";
+import { TimeLock } from "../governance/TimeLock.sol";
+import { UpgradeGovernor } from "../governance/UpgradeGovernor.sol";
 import { Proxy } from "../universal/Proxy.sol";
+import { ProxyAdmin } from "../universal/ProxyAdmin.sol";
```
Comment (Pangssu) [inline packages/contracts/contracts/test/CommonTest.t.sol:37]: ascending

File: packages/contracts/deploy/013-SecurityCouncilToken.ts [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    securityCouncilTokenProxyAddress,
+    artifact.abi,
+    hre.ethers.provider.getSigner(deployer)
+  )
+
+  // Check variable
+  assert((await token.name()) === 'KromaSecurityCouncil')
+  assert((await token.symbol()) === 'KSC')
+  assert((await token.owner()) === deployer)
+
```
Comment (Pangssu) [inline packages/contracts/deploy/013-SecurityCouncilToken.ts:32]: use `postDeployAction` of deploy options

File: packages/contracts/deploy/014-TimeLock.ts [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+    timeLockProxyAddress,
+    artifact.abi,
+    hre.ethers.provider.getSigner(deployer)
+  )
+
+  // Check variable
+  assert(
+    (await timeLock.getMinDelay()) ===
+      hre.deployConfig.proposalExecutingLatencySeconds
+  )
```
Comment (Pangssu) [inline packages/contracts/deploy/014-TimeLock.ts:39]: use postDeployAction of deploy options

File: packages/contracts/deploy/015-UpgradeGovernor.ts [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+  // Check variable
+  assert((await governor.timelock()) === timeLockProxyAddress)
+  assert((await governor.votingDelay()) === hre.deployConfig.votingDelayBlocks)
+  assert(
+    (await governor.votingPeriod()) === hre.deployConfig.votingPeriodBlocks
+  )
+  assert(
+    (await governor.proposalThreshold()) ===
+      hre.deployConfig.proposalThresholdTokens
+  )
```
Comment (Pangssu) [inline packages/contracts/deploy/015-UpgradeGovernor.ts:51]: use postDeployAction of deploy options

File: utils/chain-ops/genesis/layer_one.go [hunk 27]
```diff
@@ -254,6 +257,72 @@ func BuildL1DeveloperGenesis(config *DeployConfig) (*core.Genesis, error) {
 		return nil, err
 	}
 
+	securityCouncilTokenABI, err := bindings.SecurityCouncilTokenMetaData.GetAbi()
+	if err != nil {
+		return nil, err
+	}
+	data, err = securityCouncilTokenABI.Pack("initialize",
+		config.SecurityCouncilTokenOwner)
```
Comment (Pangssu) [inline utils/chain-ops/genesis/layer_one.go:265]:
```suggestion
data, err = securityCouncilTokenABI.Pack("initialize", config.SecurityCouncilTokenOwner)
```

**Review by Pangssu — APPROVED** [review]
